### ISSUE TITLE
Test failures in test_wsgi_unix_socket

### ISSUE THREAD

**Issue by jaraco** [cherrypy/cherrypy#1818]
When running tests on my macOS workstation, I see frequent test failures in test_wsgi_unix_socket like this:

```
―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――― WSGI_UnixSocket_Test.test_simple_request ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

func = functools.partial(<function _open_url_once at 0x10d18fdc0>, b'/', None, 'GET', None, '127.0.0.1', 54583, <cherrypy.test.test_wsgi_unix_socket.USocketHTTPConnection object at 0x10d83ca90>, 'HTTP/1.1', ssl_context=None)
cleanup = <function openURL.<locals>.on_exception at 0x10e195700>, retries = 9, trap = <class 'OSError'>

    def retry_call(func, cleanup=lambda: None, retries=0, trap=()):
        """
        Given a callable func, trap the indicated exceptions
        for up to 'retries' times, invoking cleanup on the
        exception. On the final attempt, allow any exceptions
        to propagate.
        """
        attempts = count() if retries == float('inf') else range(retries)
        for attempt in attempts:
                try:
>                       return func()

attempt    = 0
attempts   = range(0, 9)
cleanup    = <function openURL.<locals>.on_exception at 0x10e195700>
func       = functools.partial(<function _open_url_once at 0x10d18fdc0>, b'/', None, 'GET', None, '127.0.0.1', 54583, <cherrypy.test.test_wsgi_unix_socket.USocketHTTPConnection object at 0x10d83ca90>, 'HTTP/1.1', ssl_context=None)
retries    = 9
trap       = <class 'OSError'>

.tox/python/lib/python3.8/site-packages/jaraco/functools.py:323: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

url = '/', headers = [('Host', '127.0.0.1:54583')], method = 'GET', body = None, host = '127.0.0.1', port = 54583
http_conn = <cherrypy.test.test_wsgi_unix_socket.USocketHTTPConnection object at 0x10d83ca90>, protocol = 'HTTP/1.1', ssl_context = None

    def _open_url_once(
        url, headers=None, method='GET', body=None,
        host='127.0.0.1', port=8000, http_conn=http_client.HTTPConnection,
        protocol='HTTP/1.1', ssl_context=None,
    ):
        """Open the given HTTP resource and return status, headers, and body."""
        headers = cleanHeaders(headers, method, body, host, port)
    
        # Allow http_conn to be a class or an instance
        if hasattr(http_conn, 'host'):
            conn = http_conn
        else:
            kw = {}
            if ssl_context:
                kw['context'] = ssl_context
            conn = http_conn(interface(host), port, **kw)
        conn._http_vsn_str = protocol
        conn._http_vsn = int(''.join([x for x in protocol if x.isdigit()]))
        if not six.PY2 and isinstance(url, bytes):
            url = url.decode()
        conn.putrequest(
            method.upper(), url, skip_host=True,
            skip_accept_encoding=True,
        )
        for key, value in headers:
            conn.putheader(key, value.encode('Latin-1'))
        conn.endheaders()
        if body is not None:
            conn.send(body)
        # Handle response
>       response = conn.getresponse()

body       = None
conn       = <cherrypy.test.test_wsgi_unix_socket.USocketHTTPConnection object at 0x10d83ca90>
headers    = [('Host', '127.0.0.1:54583')]
host       = '127.0.0.1'
http_conn  = <cherrypy.test.test_wsgi_unix_socket.USocketHTTPConnection object at 0x10d83ca90>
key        = 'Host'
method     = 'GET'
port       = 54583
protocol   = 'HTTP/1.1'
ssl_context = None
url        = '/'
value      = '127.0.0.1:54583'

.tox/python/lib/python3.8/site-packages/cheroot/test/webtest.py:537: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <cherrypy.test.test_wsgi_unix_socket.USocketHTTPConnection object at 0x10d83ca90>

    def getresponse(self):
        """Get the response from the server.
    
        If the HTTPConnection is in the correct state, returns an
        instance of HTTPResponse or of whatever object is returned by
        the response_class variable.
    
        If a request has not been sent or if a previous response has
        not be handled, ResponseNotReady is raised.  If the HTTP
        response indicates that the connection should be closed, then
        it will be closed before the response is returned.  When the
        connection is closed, the underlying socket is closed.
        """
    
        # if a prior response has been completed, then forget about it.
        if self.__response and self.__response.isclosed():
            self.__response = None
    
        # if a prior response exists, then it must be completed (otherwise, we
        # cannot read this response's header to determine the connection-close
        # behavior)
        #
        # note: if a prior response existed, but was connection-close, then the
        # socket and response were made independent of this HTTPConnection
        # object since a new request requires that we open a whole new
        # connection
        #
        # this means the prior response had one of two states:
        #   1) will_close: this connection was reset and the prior socket and
        #                  response operate independently
        #   2) persistent: the response was retained and we await its
        #                  isclosed() status to become true.
        #
        if self.__state != _CS_REQ_SENT or self.__response:
            raise ResponseNotReady(self.__state)
    
        if self.debuglevel > 0:
            response = self.response_class(self.sock, self.debuglevel,
                                           method=self._method)
        else:
            response = self.response_class(self.sock, method=self._method)
    
        try:
            try:
>               response.begin()

response   = <http.client.HTTPResponse object at 0x10e35c9a0>
self       = <cherrypy.test.test_wsgi_unix_socket.USocketHTTPConnection object at 0x10d83ca90>

/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/http/client.py:1322: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <http.client.HTTPResponse object at 0x10e35c9a0>

    def begin(self):
        if self.headers is not None:
            # we've already started reading the response
            return
    
        # read until we get a non-100 response
        while True:
>           version, status, reason = self._read_status()

self       = <http.client.HTTPResponse object at 0x10e35c9a0>

/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/http/client.py:303: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <http.client.HTTPResponse object at 0x10e35c9a0>

    def _read_status(self):
        line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
        if len(line) > _MAXLINE:
            raise LineTooLong("status line")
        if self.debuglevel > 0:
            print("reply:", repr(line))
        if not line:
            # Presumably, the server closed the connection before
            # sending a valid response.
>           raise RemoteDisconnected("Remote end closed connection without"
                                     " response")
E           http.client.RemoteDisconnected: Remote end closed connection without response

line       = ''
self       = <http.client.HTTPResponse object at 0x10e35c9a0>

/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/http/client.py:272: RemoteDisconnected

During handling of the above exception, another exception occurred:

self = <cherrypy.test.test_wsgi_unix_socket.WSGI_UnixSocket_Test testMethod=test_simple_request>

    def test_simple_request(self):
>       self.getPage('/')

self       = <cherrypy.test.test_wsgi_unix_socket.WSGI_UnixSocket_Test testMethod=test_simple_request>

cherrypy/test/test_wsgi_unix_socket.py:81: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
cherrypy/test/helper.py:324: in getPage
    return webtest.WebCase.getPage(self, url, headers, method, body,
.tox/python/lib/python3.8/site-packages/cheroot/test/webtest.py:208: in getPage
    result = openURL(
.tox/python/lib/python3.8/site-packages/cheroot/test/webtest.py:499: in openURL
    return jaraco.functools.retry_call(
.tox/python/lib/python3.8/site-packages/jaraco/functools.py:325: in retry_call
    cleanup()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    def on_exception():
        type_, exc = sys.exc_info()[:2]
>       if isinstance(exc, raise_subcls):
E       TypeError: isinstance() arg 2 must be a type or tuple of types

exc        = RemoteDisconnected('Remote end closed connection without response')
raise_subcls = None
type_      = <class 'http.client.RemoteDisconnected'>

.tox/python/lib/python3.8/site-packages/cheroot/test/webtest.py:494: TypeError

```

If I isolate and run with `-k test_wsgi_unix_socket`, the tests pass consistently, so the issue seems to be due to interference by some other test. I've seen the failure happen twice (in a run), but most often it happens just the once.

**Comment by webknjaz** [comment]
Did it start happening after https://github.com/cherrypy/cheroot/pull/224?

**Comment by webknjaz** [comment]
You refactored `openURL()` in that PR so that if `raise_subcls` is not provided it defaults to `()` but forgot to change `WebCase.getPage(..., raise_subcls=None, ...)` which explicitly bypasses `None` and leads to that error:
```python-traceback
>       if isinstance(exc, raise_subcls):
E       TypeError: isinstance() arg 2 must be a type or tuple of types
```

https://github.com/cherrypy/cheroot/blob/52e16e7baded163418b8b02450d93af4eebedc6c/cheroot/test/webtest.py#L175-L178

This is probably only a part of the problem, of course. `RemoteDisconnected('Remote end closed connection without response')` may be related to something else, like that connection manager refactoring...

**Comment by jaraco** [comment]
I've fixed the issue in cheroot 8.2.1, [releasing here](https://travis-ci.org/cherrypy/cheroot/builds/599419689).

**Comment by jaraco** [comment]
In the above referenced commit, I've removed the duplicate signature from cherrypy CPWebCase, but I'm going to wait a while to release it, as it depends on the fix in cheroot 8.2.1.
